scispacy/candidate_generation.py

```diff
@@ -0,0 +1,307 @@
+                 verbose: bool = False) -> None:
+
+        self.ann_index = ann_index or load_approximate_nearest_neighbours_index()
+
+        self.vectorizer = tfidf_vectorizer or joblib.load(cached_path(DEFAULT_PATHS["tfidf_vectorizer"]))
+        self.ann_concept_aliases_list = ann_concept_aliases_list or \
+            json.load(open(cached_path(DEFAULT_PATHS["concept_aliases_list"])))
+
+        self.umls = umls or json.load(open(cached_path(DEFAULT_PATHS["umls_path"])))
+        self.verbose = verbose
+
+        # We need to keep around a map from text to possible canonical ids that they map to.
+        self.mention_to_concept: Dict[str, Set[str]] = defaultdict(set)
+        for concept in self.umls:
+            for alias in set(concept["aliases"]).union({concept["canonical_name"]}):
+                self.mention_to_concept[alias].add(concept["concept_id"])
+
+
+    def nmslib_knn_with_zero_vectors(self, vectors: numpy.ndarray, k: int) -> Tuple[numpy.ndarray, numpy.ndarray]:
+        """
+        ann_index.knnQueryBatch crashes if any of the vectors is all zeros.
+        This function is a wrapper around `ann_index.knnQueryBatch` that solves this problem. It works as follows:
+        - remove empty vectors from `vectors`.
+        - call `ann_index.knnQueryBatch` with the non-empty vectors only. This returns `neighbors`,
+        a list of list of neighbors. `len(neighbors)` equals the length of the non-empty vectors.
+        - extend the list `neighbors` with `None`s in place of empty vectors.
+        - return the extended list of neighbors and distances.
+        """
+        empty_vectors_boolean_flags = numpy.array(vectors.sum(axis=1) != 0).reshape(-1,)
+        empty_vectors_count = vectors.shape[0] - sum(empty_vectors_boolean_flags)
+        if self.verbose:
+            print(f'Number of empty vectors: {empty_vectors_count}')
+
+        # init extended_neighbors with a list of Nones
+        extended_neighbors = numpy.empty((len(empty_vectors_boolean_flags),), dtype=object)
+        extended_distances = numpy.empty((len(empty_vectors_boolean_flags),), dtype=object)
+
+        if vectors.shape[0] - empty_vectors_count == 0:
+            return extended_neighbors, extended_distances
+
+        # remove empty vectors before calling `ann_index.knnQueryBatch`
+        vectors = vectors[empty_vectors_boolean_flags]
+
+        # call `knnQueryBatch` to get neighbors
+        original_neighbours = self.ann_index.knnQueryBatch(vectors, k=k)
+
+        neighbors, distances = zip(*[(x[0].tolist(), x[1].tolist()) for x in original_neighbours])
+        neighbors = list(neighbors)
+        distances = list(distances)
+
+        # neighbors need to be convected to an np.array of objects instead of ndarray of dimensions len(vectors)xk
+        # Solution: add a row to `neighbors` with any length other than k. This way, calling np.array(neighbors)
+        # returns an np.array of objects
+        neighbors.append([])
```

Comment:
This is actually not the case in this particular scenario, because the objects are tuples, which means that numpy unfolds the dimension, even for objects:

```
In [16]: x = numpy.array([(1,2), (3,4)], dtype=object)

In [17]: x
Out[17]:
array([[1, 2],
       [3, 4]], dtype=object)

In [18]: x = numpy.array([(1,2), (3,4), ()], dtype=object)

In [19]: x
Out[19]: array([(1, 2), (3, 4), ()], dtype=object)

In [20]:
```

I'm sure you're right that there is a good way to fix this but I couldn't find what it is...

Reply:
seems you can tell it its a float somehow, but it looks more complicated than the current hack, so 🤷‍♂